### PR TITLE
[Review] Scene Item ID Refactor and other small changes

### PR DESCRIPTION
**Pull request by TStod**
For issues https://github.com/Palakis/obs-websocket/issues/155 , https://github.com/Palakis/obs-websocket/issues/154 , and https://github.com/Palakis/obs-websocket/issues/101

### CODE REVIEW

File: src/WSRequestHandler_Scenes.cpp [hunk 1]
```diff
@@ -74,3 +74,71 @@ void WSRequestHandler::HandleGetSceneList(WSRequestHandler* req) {
 
     req->SendOKResponse(data);
 }
+
+/**
+* Changes the order of scene items in the requested scene.
+*
+* @param {String} `scene-name (optional)` Name of the scene to reorder (defaults to current).
```
Comment (haganbmj) [inline src/WSRequestHandler_Scenes.cpp]: For some reason the documentation parsing is struggling with either the space or the parenthesis in these parameter names. I'll take a look.

I think other optionals have done this instead though
`@param {String (optional)} `scene-name` Name of the scene to reorder (defaults to current).`

Comment (TStod) [reply src/WSRequestHandler_Scenes.cpp]: Sounds good i can change that over

**Comment by TStod** [comment]
@haganbmj @Palakis wanna give this a read over?

File: src/WSRequestHandler_Sources.cpp [hunk 2]
```diff
@@ -965,6 +965,157 @@ void WSRequestHandler::HandleSetBrowserSourceProperties(WSRequestHandler* req) {
...
+ *
+ * @param {String (optional)} `scene-name` Name of the scene the source belogns to. Defaults to the current scene.
+ * @param {String} `item` Name of the source item.
+ *
+ * @api requests
+ * @name ResetSceneItem
+ * @category sources
+ * @since 4.2.0
+ */
+void WSRequestHandler::HandleResetSceneItem(WSRequestHandler* req) {
```
Comment (haganbmj) [inline src/WSRequestHandler_Sources.cpp]: This appears to be failing the build as it duplicates another method name already defined on WSRequestHandler.
```
WSRequestHandler_Sources.cpp:(.text+0x5664): multiple definition of `WSRequestHandler::HandleResetSceneItem(WSRequestHandler*)'
```

https://github.com/Palakis/obs-websocket/blob/811e3f8cc80890ba4688d941af060455480fe535/src/WSRequestHandler_SceneItems.cpp#L333-L366

Double check me, but it looks like these are indeed duplicates. You should be able to remove this definition and let it call the other definition (unless this is a better home file for this method).

File: src/WSRequestHandler_Scenes.cpp [hunk 3]
```diff
@@ -74,3 +74,71 @@ void WSRequestHandler::HandleGetSceneList(WSRequestHandler* req) {
...
+        newOrder.push_back((obs_sceneitem_t*)sceneItem);
+    }
+
+    if (obs_scene_reorder_items(obs_scene_from_source(scene), newOrder.data(), count)) {
+        req->SendOKResponse();
+    }
+    else {
+        req->SendErrorResponse("Invalid sceneItem order");
+    }
+    // Will the vector of scene items clear properly or do we have a memory leak here and on early returns?
```
Comment (TStod) [inline src/WSRequestHandler_Scenes.cpp]: @Palakis thoughts on memory leaks here?

File: src/WSRequestHandler_Sources.cpp [hunk 4]
```diff
@@ -964,6 +964,113 @@ void WSRequestHandler::HandleSetBrowserSourceProperties(WSRequestHandler* req) {
...
+    OBSSourceAutoRelease fromSource = (obs_source_t *)obs_sceneitem_get_source((obs_sceneitem_t*)referenceItem);
+    OBSSourceAutoRelease *newSource = (OBSSourceAutoRelease *)obs_source_duplicate(fromSource, obs_source_get_name(fromSource), false);
+
+    OBSSceneItemAutoRelease newItem = obs_scene_add(obs_scene_from_source(toScene), (obs_source_t *)newSource);
+    obs_sceneitem_set_visible(newItem, obs_sceneitem_visible((obs_sceneitem_t *)referenceItem));
+
+    if (!newItem) {
+        req->SendErrorResponse("Error duplicating scenee item");
+    }
+    // consider returning the new item and scene name
```
Comment (TStod) [inline src/WSRequestHandler_Sources.cpp]: thoughts on how these requests should work?

**Comment by TStod** [comment]
@Palakis if you had a moment or two to take a peek at this and let me know what you think that would be awesome

**Comment by Palakis** [comment]
@TStod apologies for the delay, gonna review this in a few hours

File: src/WSEvents.cpp [hunk 5]
```diff
@@ -757,7 +757,8 @@ void WSEvents::OnTransitionBegin(void* param, calldata_t* data) {
 /**
  * Scene items have been reordered.
  *
- * @return {String} `scene-name` Name of the scene where items have been reordered.
+ * @return {String} `name` Name of the scene where items have been reordered.
```
Comment (Palakis) [inline src/WSEvents.cpp]: Specification mismatch between this comment (`name`) and the actual code (`scene-name`, see line 774). Can't guess if it is intentional or not.

File: src/WSRequestHandler_SceneItems.cpp [hunk 6]
```diff
@@ -84,9 +88,11 @@ void WSRequestHandler::HandleGetSceneItemProperties(WSRequestHandler* req) {
 	obs_data_set_int(cropData, "top", crop.top);
 	obs_data_set_int(cropData, "right", crop.right);
 	obs_data_set_int(cropData, "bottom", crop.bottom);
-	obs_data_set_obj(data, "crop", cropData);
+	obs_data_set_obj(resItem, "crop", cropData);
 
-	obs_data_set_bool(data, "visible", obs_sceneitem_visible(sceneItem));
+  obs_data_set_bool(resItem, "visible", obs_sceneitem_visible(sceneItem));
```
Comment (Palakis) [inline src/WSRequestHandler_SceneItems.cpp]: Something wrong is happening without identation here.

File: src/WSRequestHandler_Scenes.cpp (hunk 3 above)
Comment (Palakis) [reply src/WSRequestHandler_Scenes.cpp]: I think you should release manually the obs_sceneitem_t pointers before exiting this function.

**Comment by TStod** [comment]
Sorry have been busy with work and holidays. Need to retest functions and review code again but will shoot for mid this week @Palakis

**Comment by Palakis** [comment]
@TStod thanks!

**Comment by Palakis** [comment]
By the way, I'll release a new version of obs-websocket today or tomorrow. I'll merge this after, if possible.

**Comment by TStod** [comment]
@Palakis sorry for the delay. This PR seems to be working as intended. I would like it in a newer build if possible but if we want to live on it a bit first that is ok too. IDK when you are planning nomenclature overhauls to cammelcase and such but if that is the following update I'm especially ok with waiting.

**Comment by TStod** [comment]
@Palakis I'm gonna keep testing this and will make additional PRs for bugs I find / fix

**Pull request merged**